rbac/management/utils.py

```diff
@@ -65,24 +67,48 @@ def get_principal_from_request(request):
     return get_principal(username, request, verify_principal=bool(qs_user), from_query=from_query)
 
 
-def get_principal(username, request, verify_principal=True, from_query=False):
+def get_principal(
+    username: str, request: Request, verify_principal: bool = True, from_query: bool = False
+) -> Principal:
     """Get principals from username."""
     # First check if principal exist on our side,
     # if not call BOP to check if user exist in the account.
-    tenant = request.tenant
+    tenant: Tenant = request.tenant
     try:
-        # If the username was provided through a query we must verify if it is an org admin from the BOP
+        # If the username was provided through a query we must verify if it exists in the corresponding services first.
         if from_query:
-            verify_principal_with_proxy(username, request, verify_principal=verify_principal)
+            if ITService.is_username_service_account(username=username):
+                it_service = ITService()
+                if not it_service.is_service_account_valid_by_username(
+                    user=request.user, service_account_username=username
+                ):
+                    raise serializers.ValidationError(
+                        {"detail": f"No data found for service account with username '{username}'"}
+                    )
+            else:
+                verify_principal_with_proxy(username, request, verify_principal=verify_principal)
         principal = Principal.objects.get(username__iexact=username, tenant=tenant)
     except Principal.DoesNotExist:
-        verify_principal_with_proxy(username, request, verify_principal=verify_principal)
+        # If the "from query" parameter was specified, the username was validated above, so there is no need to
+        # validate it again.
+        if not from_query:
+            if ITService.is_username_service_account(username):
+                it_service = ITService()
```

Comment:
Can we make it a global parameter or make it a singleton? To save us some time so we dont' have to init it for request

Reply:
Where would that global parameter go? I mean, the class is instantiated in the following files:

- `management/utils.py`
- `group/view.py`
- `principal/view.py`

Is there any place where we instantiate a single instance of these global objects? Otherwise I saw that I can try to wrap this into a metaclass or something and work with that, but I don't know which is the best practice here.

Thanks!

Reply:
How about this:
```
class ITService: 
    _instance = None 
    def __new__(cls, *args, **kwargs): 
        if cls._instance is None: 
            cls._instance = super().__new__(cls) 
        return cls._instance 
         
    def __init__(self, ***): 
        self.a = a 
        self.b = b
```
Then when calling ITService(), it is actually returning same instance

Reply:
TIL you could override the `__new__` method! Thank you so much for the sample! Done!